### PR TITLE
#12323 Fix types formatting in logs

### LINKED ISSUE #12323: Log formatting of classes/types is broken
**Describe the incorrect behavior you saw**
Attempt to write type in log fails.

**Describe how to cause this behavior**
Emit log message that implies some type formatting.

```
>>> from twisted.logger import eventAsText
>>> eventAsText({"log_format": "{value}", "value": RuntimeError})
"- [-#-] Unable to format event {'log_format': '{value}', 'value': <class 'RuntimeError'>}: object.__format__() takes exactly one argument (0 given)"
```

**Describe the correct behavior you'd like to see**
Type is being formatted just like in `format()` call.

**Additional context**
The regression was introduced in Twisted 24.3.0 (https://github.com/twisted/twisted/pull/12040)
23.10.0 works fine:
```
>>> from twisted.logger import eventAsText
>>> eventAsText({"log_format": "{value}", "value": RuntimeError})
"- [-#-] <class 'RuntimeError'>"
```

### PR DESCRIPTION
**Pull request by thevintik**
Fix PotentialCallWrapper and add tests

## Scope and purpose

Fixes #12323

### CODE REVIEW

**Comment by adiroiban** [comment]
Hi Igor. Thanks for the PR.

Is this ready to review ?
Are you happy with the current code in the PR and would like it to be merged ?

Cheers

**Comment by thevintik** [comment]
Hi Adi. Yes, it is ready. No more changes are intended.

**Comment by thevintik** [comment]
please review

File: src/twisted/logger/test/test_format.py [hunk 1]
```diff
@@ -95,6 +95,14 @@ def where(self) -> str:
             "hello world", self.format("hello {what.where()}", what=World())
         )
 
+    def test_formatTypes(self) -> None:
+        """
+        Classes/types formatting.
```
Comment (adiroiban) [inline src/twisted/logger/test/test_format.py]:
```suggestion
        C{formatEvent} can handle formats that have classes or types as input.
```

File: src/twisted/newsfragments/12323.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Formatting types in logs no more results in "Unable to format event" message (was broken since 24.3.0).
```
Comment (adiroiban) [inline src/twisted/newsfragments/12323.bugfix]: I am not very familiar with the `twisted.logger` API.

As part of our best practices, the release notes fragments should include the full name of the public API that is affected by this change.

I have added 2 API names, but I guess that this change is affect other public APIs

```suggestion
twisted.logger.formatEvent and twisted.logger.eventAsText can now format the output having types/classes as input.
This was a regression introduce in Twisted 24.3.0.
```

File: src/twisted/logger/_format.py [hunk 3]
```diff
@@ -208,13 +208,13 @@ def __getitem__(self, name: str) -> object:
         return PotentialCallWrapper(value)
 
     def __format__(self, format_spec: str) -> str:
-        return self._wrapped.__format__(format_spec)
+        return format(self._wrapped, format_spec)
 
     def __repr__(self) -> str:
-        return self._wrapped.__repr__()
+        return repr(self._wrapped)
```
Comment (adiroiban) [inline src/twisted/logger/_format.py:214]: just asking ... same question for `__str__`

Are these changes really needed ?

do we have a new test for this change ?

File: src/twisted/logger/test/test_format.py [hunk 4]
```diff
@@ -95,6 +95,14 @@ def where(self) -> str:
             "hello world", self.format("hello {what.where()}", what=World())
         )
 
+    def test_formatTypes(self) -> None:
```
Comment (adiroiban) [inline src/twisted/logger/test/test_format.py]: I think that is best to keep `formatEvent` in the name, to highlight that this is a test for the `t.logger.formatEvent` public API.
 
Can was also have a test for `twisted.logger.eventAsText` which is the failing API in the initial bug report?

```suggestion
    def test_formatEventTypes(self) -> None:
```

**Review by adiroiban — APPROVED** [review]
Thanks for the PR.

Changes looks good.

I tried to improve the text for the release notes.
It is not clear which API is affected / fixed by this PR.

I have a minor comment in terms if test coverage... but not a blocker.

Thanks again for the fix

File: src/twisted/logger/_format.py (hunk 3 above)
Comment (thevintik) [reply src/twisted/logger/_format.py:214]: For the very specific issue we encountered with, only changing of `__format__` is enough.
I think I'll add tests for `__getattr__` as well.
As for `repr` and `str` - I'm not sure that they are ever used here, and can't come up with tests. So they are changed just for consistence. But probably they can be completely removed.

File: src/twisted/newsfragments/12323.bugfix (hunk 2 above)
Comment (thevintik) [reply src/twisted/newsfragments/12323.bugfix]: Actually, `formatEventAsClassicLogText`, `textFileLogObserver` are also affected. `eventAsText` is the very basic one which reused by others. So it was specified in the issue as simplest way to reproduce.
Do I need to specify and test them all? Or `eventAsText` will be enough?

File: src/twisted/logger/test/test_format.py (hunk 4 above)
Comment (thevintik) [reply src/twisted/logger/test/test_format.py]: I changed `formatEvent` test to `eventAsText` one. Not sure if we should add all higher-level APIs tests then.

File: src/twisted/logger/_format.py (hunk 3 above)
Comment (adiroiban) [reply src/twisted/logger/_format.py:214]: I think that we can keep them.

We have at least one test that covers this code...and the test seems to pass

So we are ok.

Thanks

File: src/twisted/newsfragments/12323.bugfix (hunk 2 above)
Comment (adiroiban) [reply src/twisted/newsfragments/12323.bugfix]: eventAsText should be enough. thanks

**Comment by adiroiban** [comment]
Many thanks for your followup. All looks good and I have merged the PR.
Great work!
